### DISCUSSION TITLE
Automap fails on mysql 8 when user does not have rights on all foreign key columns

### DISCUSSION THREAD

**Discussion by ellepdesk** [sqlalchemy/sqlalchemy#12081] [Usage Questions] [answered]
After updating our production db to mysql 8 we found some of our code did no longer work.
related to #11967

It seems the function `_correct_for_mysql_bugs_88718_96365` cannot find certain columns if the current user had no rights to view them and raises an error. The user has limited access to the DB, and had no need for these rights before.

For now I've fixed it by allowing select on the required columns, but it feels unneeded.
adding `engine.connect().close()
engine.dialect._needs_correct_for_88718_96365 = False`
Also works around the issue.

I think this would better be served as a warning, something like 
`
                try:
                    fkey["referred_table"] = rec["TABLENAME"]
                except KeyError:
                    self.logging.warning(f"unable to verify foreign key relation(s), cannot inspect: {fkey['referred_table']} {fkey['referred_columns']}" )
                    continue
`
Or at least a better error message.

`[ERROR] KeyError: 'TABLENAME'
Traceback (most recent call last):
  File "/var/lang/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/var/task/lambda_function.py", line 26, in <module>
    dbc = BoilerDBC(db_url=mysql_db_url, enable_views=False)
  File "/var/task/boiler_tools/BoilerDBC.py", line 1109, in __init__
    super().__init__(db_url=db_url, session_type=BoilerDBSession, views=views, connect_args=connect_kwargs)
  File "/var/task/boiler_tools/peeekslib/databaseconnector.py", line 88, in __init__
    metadata = self._create_metadata(tables, views)
  File "/var/task/boiler_tools/peeekslib/databaseconnector.py", line 124, in _create_metadata
    metadata.reflect(self.engine)
  File "/var/task/sqlalchemy/sql/schema.py", line 4901, in reflect
    Table(name, self, **reflect_opts)
  File "<string>", line 2, in __new__
  File "/var/task/sqlalchemy/util/deprecations.py", line 375, in warned
    return fn(*args, **kwargs)
  File "/var/task/sqlalchemy/sql/schema.py", line 618, in __new__
    with util.safe_reraise():
  File "/var/task/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/var/task/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/var/task/sqlalchemy/sql/schema.py", line 614, in __new__
    table._init(name, metadata, *args, **kw)
  File "/var/task/sqlalchemy/sql/schema.py", line 689, in _init
    self._autoload(
  File "/var/task/sqlalchemy/sql/schema.py", line 724, in _autoload
    conn_insp.reflect_table(
  File "/var/task/sqlalchemy/engine/reflection.py", line 795, in reflect_table
    self._reflect_fk(
  File "/var/task/sqlalchemy/engine/reflection.py", line 950, in _reflect_fk
    fkeys = self.get_foreign_keys(
  File "/var/task/sqlalchemy/engine/reflection.py", line 564, in get_foreign_keys
    return self.dialect.get_foreign_keys(
  File "<string>", line 2, in get_foreign_keys
  File "/var/task/sqlalchemy/engine/reflection.py", line 55, in cache
    ret = fn(self, con, *args, **kw)
  File "/var/task/sqlalchemy/dialects/mysql/base.py", line 2891, in get_foreign_keys
    self._correct_for_mysql_bugs_88718_96365(fkeys, connection)
  File "/var/task/sqlalchemy/dialects/mysql/base.py", line 2970, in _correct_for_mysql_bugs_88718_96365
    fkey["referred_table"] = rec["TABLENAME"]`

**Comment by CaselIT** [comment]
Hi,

The warning seems like a good compromise. A PR to that effect would be accepted. Any opinion on this @zzzeek ?

**Reply by zzzeek** [reply]
I think it's highly unusual for MySQL user accounts to have limited permissions on schema introspection tables, and especially if someone wants to write a program that introspects schema level information from the database, it is entirely appropriate that extremely basic tables like `information_schema.columns` be visible to the account.     MySQL is unusual in that most of our schema introspection is done by regexing "SHOW CREATE TABLE" but if this were any other database, it would be a given that the information_schema / catalog tables need to be visible to the user.

so I'd be -1 on any change here, certainly not any kind of "warn and instead do broken reflection", we can document that information_schema needs to be visible to the user account for table reflection to work, but I would think this would just be a note in the tutorial / basic docs for reflection "note: reflection  requires full viewable rights to system catalog tables like pg_catalog /inforamtion_schema etc."

Note this is not local to the change in a98d31621d58d45e2a4d74351282cedddcf since we didnt change the tables being queried.

**Reply by ellepdesk** [reply]
I don't think I can explicitly allow access to information_schema, I don't know if this is an Amazon RDS thing or a mysql thing.

The query succeeds, but it only includes columns I have select access on. Previously this was not needed, the user is only used to select data so foreign key checks were never needed.

So my problem still stands that unless I grant at least select on all columns used as foreign key, the user cannot reflect the db structure (by throwing a low-level python error).

**Reply by zzzeek** [reply, marked as answer]
> I don't think I can explicitly allow access to information_schema, I don't know if this is an Amazon RDS thing or a mysql thing.

I dont know amazon RDS but on a vanilla MySQL install, information_schema privs are defaulted, allowing a user to access all rows that apply to tables to which they have access, see https://dev.mysql.com/doc/mysql-infoschema-excerpt/8.0/en/information-schema-introduction.html#information-schema-privileges  

column-level privs are very unusual, apparently mysql can do this but this is not a thing I've ever seen used before

> The query succeeds, but it only includes columns I have select access on. Previously this was not needed, the user is only used to select data so foreign key checks were never needed.

you are using automap, the purpose of which is to introspect the structure of tables in the database.   if your program relies on this feature, then of course you have to ensure database logins have privileges to the schema tables that store this information.    that some implementation detail of this changed in SQLAlchemy is not an argument against that.   if you cannot grant permissions to users then you should use explicit mappings in source code, which is more efficient and accurate anyway.  if you want to generate these ahead of time you can try out a tool like [sqlacodegen](https://github.com/agronholm/sqlacodegen)

**Reply by ellepdesk** [reply]
I'll have a look at using explicit mappings, for now it's up and running again with more rights and/or disabling the check completely.

**Reply by CaselIT** [reply]
your reasoning makes sense to me mike.

I guess the only change that could be done is to raise a better error in this case?

**Comment by honglei** [comment]
Thanks for the clarification.

I’d like to approach this from a slightly different angle.

Granting different users access to only a subset of tables is a very common and standard practice in MySQL. Table-level privileges are widely used in real systems (e.g. multi-tenant setups, data governance, analytics access control, etc.).

In such setups, it is expected that:
- a user can access only part of the schema
- INFORMATION_SCHEMA automatically reflects this by exposing only metadata for accessible objects

So from the database perspective, “partial schema visibility” is not an edge case, but a normal operating mode.

Given that, it seems reasonable that SQLAlchemy reflection should behave consistently with that model:

→ reflect whatever metadata is visible to the current user  
→ ignore or gracefully skip objects that are not accessible  

In the current behavior:

- SQLAlchemy is able to enumerate tables and parse FK definitions via SHOW CREATE TABLE
- but the MySQL FK correction step depends on INFORMATION_SCHEMA
- when INFORMATION_SCHEMA returns no rows (due to privilege filtering), the code assumes the result exists and raises:

    KeyError('TABLENAME')

This effectively causes reflection to fail entirely, even though partial metadata is already available.

From a usability perspective, this makes reflection unusable in a valid and common privilege configuration.

A more robust behavior would be:

- proceed with the metadata that is available
- skip FK correction when INFORMATION_SCHEMA lookup returns empty
- optionally emit a warning

This would align SQLAlchemy’s behavior with MySQL’s privilege model:

→ “return what is visible, ignore what is not”

rather than failing due to internal assumptions.

Importantly, this does not require full support for restricted introspection — only that reflection degrades gracefully instead of aborting.

Thanks for considering this perspective.

**Reply by zzzeek** [reply]
> Thanks for the clarification.
> 
> I’d like to approach this from a slightly different angle.
> 
> Granting different users access to only a subset of tables is a very common and standard practice in MySQL. Table-level privileges are widely used in real systems (e.g. multi-tenant setups, data governance, analytics access control, etc.).
> 
> In such setups, it is expected that:
> 
>     * a user can access only part of the schema
> 
>     * INFORMATION_SCHEMA automatically reflects this by exposing only metadata for accessible objects
> 
> 
> So from the database perspective, “partial schema visibility” is not an edge case, but a normal operating mode.

sure and this agrees with what we see at https://dev.mysql.com/doc/mysql-infoschema-excerpt/8.0/en/information-schema-introduction.html#information-schema-privileges 


> 
> → reflect whatever metadata is visible to the current user → ignore or gracefully skip objects that are not accessible
> 
> In the current behavior:
> 
>     * SQLAlchemy is able to enumerate tables and parse FK definitions via SHOW CREATE TABLE
> 
>     * but the MySQL FK correction step depends on INFORMATION_SCHEMA
> 
>     * when INFORMATION_SCHEMA returns no rows (due to privilege filtering), the code assumes the result exists and raises:
>       KeyError('TABLENAME')

we didnt dig deeply into the original problem here, the phrase "The query succeeds, but it only includes columns I have select access on. " led me to believe Amazon RDS had some kind of column-level rights in play.

the issue reported here is not that the table threw an error on SELECT it's just that it isnt giving back privileged rows when we are trying to do a case correction.   then again we dont know if there are further issues with automap subsequent (doesnt automap want to make a whole class mapping for that remote table?)

I would want to get an update on the status of bugs         https://bugs.mysql.com/bug.php?id=88718 and  https://bugs.mysql.com/bug.php?id=96365 to see where we are at and I'd want to see a real POC on what the issue specifically is before we decide to ignore missing rows.   would be preferable to not require this "fix" step anymore if that's an option.

**Reply by honglei** [reply]
Environment:
- MySQL 8.0.32 (Community Server)
- SQLAlchemy 2.0.49

Setup:
- Table `issue_problem_image` has FK → `issue_problem(id)`
- The current user has NO SELECT privilege on both tables
- SHOW CREATE TABLE works
- INFORMATION_SCHEMA.COLUMNS returns 0 rows for these tables(expected under privilege filtering)

Observed behavior:

- get_foreign_keys() already returns correct FK metadata before the correction step

- With correction enabled (_needs_correct_for_88718_96365 = True):
    → triggers INFORMATION_SCHEMA.COLUMNS lookup
    → returns empty result set
    → leads to KeyError('TABLENAME')

- With correction disabled:
    engine.dialect._needs_correct_for_88718_96365 = False
    → reflection works normally

Conclusion on MySQL 8.0.32:

→ FK metadata is correct without the correction step  
→ the workaround for 88718 / 96365 does not appear necessary  
→ INFORMATION_SCHEMA returning empty is expected behavior under privilege filtering  
→ the correction step introduces failure in this case  

Given this, would it be reasonable to:

- skip the correction step when INFORMATION_SCHEMA lookup returns empty  
- or retain the already parsed FK metadata instead of assuming the lookup must succeed  

Also, this aligns with the idea that avoiding this fix step entirely would be preferable if it's no longer needed for modern MySQL versions.

**Reply by zzzeek** [reply]
the workarounds are needed for MySQL versions between 8.0.1 and 8.0.13, and the 96365 workaround is needed for OSX.

the bigger issue here is that fixing all of this wont really help you because the remote table is still not reflectable, you would need to use resolve_fks=False as well.

bigger than that is uncovering potential mis-use of the UnreflectableTableError where the primary table is blocked when it's really the remote table that has the problem, and if confirmed that should be improved, and this is cross-dialect.

**Reply by zzzeek** [reply]
OK whatever we do for _correct_for_mysql_bugs_88718_96365 does not actually matter here, because automap still cannot be used under these conditions.

using **mariadb**, which does not have the need for _correct_for_mysql_bugs_88718_96365, start with this SQL script:

```sql
-- Run as root: mysql -u root < poc_setup.sql

DROP DATABASE IF EXISTS fk_reflect_test;
CREATE DATABASE fk_reflect_test;
USE fk_reflect_test;

CREATE TABLE parent_table (
    id INTEGER PRIMARY KEY,
    data VARCHAR(50)
) ENGINE=InnoDB;

CREATE TABLE child_table (
    id INTEGER PRIMARY KEY,
    parent_id INTEGER,
    info VARCHAR(50),
    CONSTRAINT fk_parent FOREIGN KEY (parent_id)
        REFERENCES parent_table(id)
) ENGINE=InnoDB;

INSERT INTO parent_table VALUES (1, 'parent row');
INSERT INTO child_table VALUES (1, 1, 'child row');

DROP USER IF EXISTS 'fk_reflect_user'@'%';
CREATE USER 'fk_reflect_user'@'%' IDENTIFIED BY 'testpass';

-- user can see child_table but NOT parent_table
GRANT SELECT ON fk_reflect_test.child_table
    TO 'fk_reflect_user'@'%';

FLUSH PRIVILEGES;

```

then test case:

```py
"""POC: automap prepare() fails when FK targets are inaccessible.

Setup:  mysql -u root < poc_setup.sql
Run:    python poc_reflect.py
"""

from sqlalchemy import create_engine, inspect
from sqlalchemy.ext.automap import automap_base

url = (
    "mysql+mysqldb://"
    "fk_reflect_user:testpass@localhost/fk_reflect_test"
)
engine = create_engine(url, echo=False)

insp = inspect(engine)
print("get_table_names():", insp.get_table_names())
print("child_table FKs:", insp.get_foreign_keys("child_table"))

Base = automap_base()
Base.prepare(autoload_with=engine, reflection_options=dict(resolve_fks=False))

print("mapped classes:", Base.classes.keys())
print("metadata.tables:", list(Base.metadata.tables.keys()))

```

note the resolve_fks=False directive, indicating we should skip trying to load up parent_table.   the operation still fails, at the automap level:

```
$ python poc_reflect.py 
get_table_names(): ['child_table']
child_table FKs: [{'name': 'fk_parent', 'constrained_columns': ['parent_id'], 'referred_schema': None, 'referred_table': 'parent_table', 'referred_columns': ['id'], 'options': {}}]
Traceback (most recent call last):
  File "/home/classic/dev/sqlalchemy/poc_reflect.py", line 21, in <module>
    Base.prepare(autoload_with=engine, reflection_options=dict(resolve_fks=False))
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 2, in prepare
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/util/deprecations.py", line 281, in warned
    return fn(*args, **kwargs)  # type: ignore[no-any-return]
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/ext/automap.py", line 1353, in prepare
    _relationships_for_fks(
    ~~~~~~~~~~~~~~~~~~~~~~^
        automap_base,
        ^^^^^^^^^^^^^
    ...<5 lines>...
        generate_relationship,
        ^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/ext/automap.py", line 1513, in _relationships_for_fks
    referred_table = fks[0].column.table
                     ^^^^^^^^^^^^^
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/util/langhelpers.py", line 1213, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
                                           ~~~~~~~~~^^^^^
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/sql/schema.py", line 3505, in column
    return self._resolve_column()
           ~~~~~~~~~~~~~~~~~~~~^^
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/sql/schema.py", line 3530, in _resolve_column
    raise exc.NoReferencedTableError(
    ...<5 lines>...
    )
sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'child_table.parent_id' could not find table 'parent_table' with which to generate a foreign key to target column 'id'

```

so we can talk about _correct_for_mysql_bugs_88718_96365 not being great, we can talk about error messages that may want to encourage the use of resolve_fks=False, however, for the purposes of making a real issue to fix a real use case, that's not defined yet.  What should automap be doing here?   it seems like this is really a feature request for automap, a parameter to prepare() that is something like "skip_unreflectable_related_tables", something like that.

**Reply by honglei** [reply]
Automap under restricted privileges is a valid separate topic. However, the immediate issue exposed by our actual DB-GPT usage is narrower: DB-GPT relies on SQLAlchemy reflection/introspection rather than automap-style ORM generation, and in our case the inaccessible tables are not required for the application’s actual work. So our first priority is that reflection should degrade gracefully under restricted privileges, instead of failing hard when some FK-related metadata is not visible. More generally, under restricted database privileges, the expected behavior should not be a reconstructed privileged global schema view, but correct behavior within the authorized subset: accessible tables should still reflect correctly, and inaccessible related tables or FK targets should be treated as unresolved or skipped rather than causing the whole reflection flow to fail. For automap, if a parent table is not accessible and therefore cannot be mapped, the natural fallback would be to skip generating the corresponding relationship instead of treating that as a fatal error.

**Reply by zzzeek** [reply]
this is the funny thing about github discussions, this discussion was opened by someone trying to get automap to work.   the use case was murky (why use automap if you can't reflect the table that you need to make relationships).    you're someone else who is focused on a smaller part of the problem.

the current contract of Table(...autoload_with=...) is that it follows out foreign keys to those related tables.  if you specify `resolve_fks=False` then it does not do that, but it still produces ForeignKey() with the name of the related table.

In this case, it's not controversial to limit the scope of _correct_for_mysql_bugs_88718_96365() to do the 88718 part only on MySQL 8.0.1-8.0.13 and do to the 96365 part only on OSX where lower_case_table_names=2.    If we modified the method to work in that way, it would no longer impact anyone using MySQL 8.0.14+ on linux.  an OSX setup is vastly likely to be a developer workstation which would more commonly have open privileges.

for the change where we have _correct_for_mysql_bugs_88718_96365() run in all MySQL 8 cases as it does now and just silently degrade (or with a warning, which is often not much better than silent) is more controversial because that's asking reflection to deliver the wrong answer on MySQL 8.0.1-8.0.13 or OSX.  it's not that it just skips getting additional information, it's that the ForeignKey() will have the *wrong* information in it - the name will be wrong.    

I think we would be better off further limiting the scope of _correct_for_mysql_bugs_88718_96365() to only those platforms affected (which is how it was actually written; it only takes effect for MySQL 8 and not MariaDB or older MySQL versions; at the time, neither issue had been fixed in MySQL)


tldr this is the non-controversial patch, test if this resolves your real world issues

```diff
diff --git a/lib/sqlalchemy/dialects/mysql/base.py b/lib/sqlalchemy/dialects/mysql/base.py
index bc8875efc..ff177e538 100644
--- a/lib/sqlalchemy/dialects/mysql/base.py
+++ b/lib/sqlalchemy/dialects/mysql/base.py
@@ -3187,7 +3187,12 @@ class MySQLDialect(default.DefaultDialect):
         )
 
         self._needs_correct_for_88718_96365 = (
-            not self.is_mariadb and self.server_version_info >= (8,)
+            not self.is_mariadb
+            and self.server_version_info >= (8,)
+            and (
+                self.server_version_info < (8, 0, 14)  # bug 88718
+                or self._casing == 2  # bug 96365, macOS
+            )
         )
 
         self.delete_returning = (

```

**Reply by zzzeek** [reply]
also noting you need to put resolve_fks=False in your reflection commands

**Reply by honglei** [reply]
The patch for `_needs_correct_for_88718_96365`  resolved my issue. Thanks!
